packages/liveblocks-client/src/immutable.ts

```diff
@@ -271,33 +272,48 @@ function getParentsPath(node: LiveNode): Array<string | number> {
   return path;
 }
 
-export function patchImmutableObject<T>(state: T, updates: StorageUpdate[]): T {
+export function patchImmutableObject<S extends JsonObject>(
+  state: S,
+  updates: StorageUpdate[]
+): S {
   return updates.reduce(
     (state, update) => patchImmutableObjectWithUpdate(state, update),
     state
   );
 }
 
-function patchImmutableObjectWithUpdate<T>(state: T, update: StorageUpdate): T {
+function patchImmutableObjectWithUpdate<S extends JsonObject>(
+  state: S,
+  update: StorageUpdate
+): S {
   const path = getParentsPath(update.node);
   return patchImmutableNode(state, path, update);
 }
 
-function patchImmutableNode(
-  state: any,
+function patchImmutableNode<S extends Json>(
+  state: S,
   path: Array<string | number>,
   update: StorageUpdate
-): any {
+): S {
+  // FIXME: Split this function up into a few smaller ones! In each of them,
+  // the types can be define much more narrowly and correctly, and there will
+  // be less type shoehorning necessary.
```

Comment:
Leaving this as a todo for later. Doing this will make most of the type hacking in here unnecessary.